### PR TITLE
feat: Add typing support

### LINKED ISSUE #31: Support typing
> [!NOTE]
> In the meantime, type stubs are installable using typeshed for nanoid v2.0.0.
>
> `pip install types-nanoid`
>
> [Related typeshed PR](https://github.com/python/typeshed/pull/12074)

Types are increasingly popular for modern python projects:
- [PEP484](https://peps.python.org/pep-0484/)
- [PEP561](https://peps.python.org/pep-0561/)
- etc

It would be great to remove nanoid from mypy overrides ignore missing imports in my projects:)

### PR DESCRIPTION
**Pull request by oliverlambson**
> [!NOTE]
> In the meantime, type stubs are installable using typeshed for nanoid v2.0.0.
>
> `pip install types-nanoid`
>
> [Related typeshed PR](https://github.com/python/typeshed/pull/12074)

Closes #31 

- Keeps support for python 3.6 (since that's what the CI is running on, I assume that is a target version even though 3.6 & 3.7 are EOL)
- Adds `mypy` as a dev dependency
- Adds type annotations to module and tests, passing `mypy --strict`
- Adds `py.typed` marker so types are recognised when installing package

### CODE REVIEW

**Comment by oliverlambson** [comment]
@puyuan I know this isn't a very active repo, but would appreciate the review here since quite a lot of people depend on this project with [over 600k downloads last month (nice!)](https://pypistats.org/packages/nanoid)—I think having types would be quite a nice quality of life improvement

Perhaps you could also add @aidarkhanov as a maintainer (if they're interested)? I'd also be happy to contribute if needed

**Comment by telesvar** [comment]
> @puyuan I know this isn't a very active repo, but would appreciate the review here since quite a lot of people depend on this project with [over 600k downloads last month (nice!)](https://pypistats.org/packages/nanoid)—I think having types would be quite a nice quality of life improvement
> 
> Perhaps you could also add @aidarkhanov as a maintainer (if they're interested)? I'd also be happy to contribute if needed

Let me take a look, maybe I can approve the PR

PS. no, I can't merge your PR, cause I don't have write access :(

**Review by telesvar — APPROVED** [review]
LGTM!

**Comment by oliverlambson** [comment]
@aidarkhanov in the meantime, I've [added stubs to typeshed](https://github.com/python/typeshed/pull/12074). That allows for types without updating this package at all, just `pip install types-nanoid` 🤞

File: nanoid/method.py [hunk 1]
```diff
@@ -2,17 +2,22 @@
 from __future__ import division
 
 from math import ceil, log
+from typing import Callable, Sequence
+from typing_extensions import TypeAlias
```
Comment (Avasam) [inline nanoid/method.py:6]: If you add `typing_extensions` at runtime, you should also add it to the project's dependencies!
Type-checkers special-case `typing_extensions`, but it's not part of the stdlib !

You can avoid using it at runtime by doing:
```suggestion
from __future__ import annotations # move this to the top of the file

from typing import Callable, Sequence, TYPE_CHECKING
if TYPE_CHECKING:
	from typing_extensions import TypeAlias
```

or
```py
from typing import Callable, Sequence, TYPE_CHECKING
if TYPE_CHECKING:
	from typing_extensions import TypeAlias

	Algorithm: TypeAlias = Callable[[int], Sequence[int]]
else:
	Algorithm = Callable[[int], Sequence[int]]
```

File: setup.py [hunk 2]
```diff
@@ -1,11 +1,11 @@
-from setuptools import setup
+from setuptools import setup  # type: ignore
```
Comment (Avasam) [inline setup.py:1]: You can add `types-setuptools` in your dev requirements to avoid this typing issue

**Comment by Avasam** [comment]
Your `py.typed` file is misplaced, it should go in the `nanoid` folder if you want it to be picked-up automatically by [setuptools >=69](https://setuptools.pypa.io/en/latest/history.html#v69-0-0)

For the marker to be picked up by setuptools < 69 (or other build backends), you also need to specify `package_data` in your setup file to include the `py.typed` marker.
See https://peps.python.org/pep-0561/#packaging-type-information

**Comment by oliverlambson** [comment]
> Your `py.typed` file is misplaced, it should go in the `nanoid` folder if you want it to be picked-up automatically by [setuptools >=69](https://setuptools.pypa.io/en/latest/history.html#v69-0-0)
> 
> For the marker to be picked up by setuptools < 69 (or other build backends), you also need to specify `package_data` in your setup file to include the `py.typed` marker. See https://peps.python.org/pep-0561/#packaging-type-information

@Avasam thanks for the helpful comments, I've actually given up on this PR as the repo owner has disappeared. I've published a rust-backed implementation that can act as a drop-in replacement: [fastnanoid](https://pypi.org/project/fastnanoid/). It's 3x faster and has type-annotations.

```diff
- from nanoid import generate
+ from fastnanoid import generate
```

**Pull request closed without merging**